DESIGN.md

```diff
@@ -28,6 +28,196 @@ daemon is spawned to handle future requests in its stead.
+
+Version-specific deletions involve a few steps. A deletion targeting the latest
+version of an object has to:
+
+- delete the specified version identified by `versionId`
+- replace the master version with a place holder for deletion version
+- initiate a repair operation to update the value of the master version:
+  - involves listing the versions of the object and get the latest
+  version to replace the place holder delete version
+  - if no more versions exists, metadata deletes the master version
+
+##### GET
+
+-  no options: original GET operation, will get the master version
+- `versionId: <versionId>` retrieve a specific version
+
+The implementation of a GET operation does not change compared to the standard
+version. A standard GET without versioning information would get the master
+version of a key. A version-specific GET would retrieve the specific version
+identified by the key for that version.
+
+##### LIST
+
+For a standard LIST on a bucket, metadata iterates through the keys by using
+the separator `.` as an extra delimiter. For a listing of all versions of a
+bucket, there is no change compared to the original listing function. Instead,
+the API component returns all the keys in a List Objects call and filters for
+just the keys of the master versions in a List Object Versions call.
+
+For example, a standard LIST operation against the keys in a table below would
+return from metadata the list of `[ foo/bar, bar, qux/quz, quz ]` using the
+separator represented by `.` as the extra delimiter.
+
+| key |
+|-----|
+| foo/bar |
+| foo/bar.v2 |
+| foo/bar.v1 |
+| bar |
+| qux/quz |
+| qux/quz.v2 |
+| qux/quz.v1 |
+| quz |
+| quz.v2 |
+| quz.v1 |
+
+#### Delete Markers
+
+In buckets that have been configured for versioning, the Delete Object API
+action will not result in the deletion of an object or version unless a specific
+version ID is provided in the request query, e.g. `DELETE /foo?versionId=v1`.
+
+If no version ID is provided, a special version called a delete marker is
+created. When performing actions upon an object whose latest version is a delete
+maker, the S3 API will return a 404 NoSuchKey error and behave as if that object
+has been deleted. To restore a previous version as the latest version of an
+object, the delete marker must be deleted, by the same process as deleting any
+other version.
+
+The metadata engine's versioning processing is agnostic to whether a specific
+version is a delete marker. The S3 API stores information about whether
+a version is a delete marker in its metadata and performs a check to retrieve
+that information.
+
+#### Null Version Mapping
+
+Null versions are created when versioning has not been enabled or when
+versioning is suspended. Previous null versions are overwritten by new null
+versions. Within metadata, this means the null version is stored by overwriting
+the master version only, instead of overwriting the master version and
+generating a new version.
```

Comment:
Expand to explain that the null version need not be the master version for example when versioning is enabled and new versions are put, the null version will just be another null version.

Also, when versioning is suspended and you have a null version that is not the master, I assume the non-master null version gets overwritten regardless?  Please specify.